core/cloudflow-spark/src/main/scala/cloudflow/spark/SparkStreamlet.scala

```diff
@@ -84,60 +80,69 @@ trait SparkStreamlet extends Streamlet[SparkStreamletContext] with Serializable
   protected def createLogic(): SparkStreamletLogic
 
   override final def run(context: SparkStreamletContext): StreamletExecution = {
+    val completionPromise = Promise[Dun]()
+    val completionFuture  = completionPromise.future
+
+    // these values are candidates to move to the configuration system
     val InitialDelay                 = 2 seconds
     val MonitorFrequency             = 5 seconds
     implicit val system: ActorSystem = ActorSystem("spark_streamlet", context.config)
 
-    readyPromise.trySuccess(Dun)
     val streamletQueryExecution = createLogic.buildStreamingQueries
 
     new StreamletExecution {
-      val readyFuture = readyPromise.future
 
-      // schedule a function to check periodically if any of the queries
-      // raised an exception
-      val done = system.scheduler.schedule(InitialDelay, MonitorFrequency) {
-        val failed = streamletQueryExecution.queries.filter(_.exception.nonEmpty)
-        if (failed.nonEmpty) {
-          // if any of the queries has an exception, stop them all
-          val errors = failed.map(f ⇒ f.name -> f.status.message).mkString(", ")
-          system.log.error(s"Queries failed. Stopping Process. Reason: $errors")
-
-          // stop all query execution
-          streamletQueryExecution.stop()
-
-        } else ()
+      // schedule a function to check periodically if any of the queries stopped
+      val scheduledQueryCheck: Cancellable = system.scheduler.schedule(InitialDelay, MonitorFrequency) {
+        val someQueryStopped = streamletQueryExecution.queries.exists(!_.isActive)
+        if (someQueryStopped) completionPromise.completeWith(stop())
       }
 
       // this future will be successful when any of the queries face an exception
       // or is stopped. The runner needs to await on this future and exit only when it
       // succeeds
-      def completed: Future[Dun] = {
-        if (streamletQueryExecution.queries.forall(!_.isActive)) {
-          // not active means the queries have either been stopped or they faced an exception
-          done.cancel()
-
-          val exceptions = streamletQueryExecution.queries.flatMap(_.exception.map(_.cause).toList)
-          if (exceptions.nonEmpty) {
-            // fail the future with a list of exceptions returned by the queries
-            val _ = completionPromise.tryFailure(ExceptionAcc(exceptions))
-          } else {
-            // succeed the future and we can now go and stop the runner
-            // Note: we still now have no way to restart any query
-            val _ = completionPromise.trySuccess(Dun)
+      def completed: Future[Dun] = completionFuture
+      val ready: Future[Dun]     = Future.successful(Dun)
+      def stop(): Future[Dun] = {
+        streamletQueryExecution.stop()
+        scheduledQueryCheck.cancel()
+        poll(streamletQueryExecution.queries.forall(!_.isActive), 1.second, StopTimeout, system.scheduler)
```

Comment:
FYI there is a cancellable scheduler as well. you could pass in a promise into the  runnable and complete it with failure when it times out or complete with result if it succeeds. Might simplify this code a bit.